### ISSUE TITLE
The implementation of PackageInstaller on Xiaomi devices is buggy

### ISSUE THREAD

**Issue by PsykoRider** [Tobi823/ffupdater#150]
Stacktrace:
```
java.lang.IllegalArgumentException: File '/storage/emulated/0/Android/data/de.marmaro.krt.ffupdater/files/Download/com.brave.browser.apk' must exists.
	at de.marmaro.krt.ffupdater.security.FingerprintValidator.checkApkFile(Unknown Source:87)
	at de.marmaro.krt.ffupdater.installer.SecureAppInstaller.secureInstallAsync(Unknown Source:116)
	at de.marmaro.krt.ffupdater.installer.SecureAppInstaller.access$secureInstallAsync(Unknown Source:0)
	at de.marmaro.krt.ffupdater.installer.SecureAppInstaller$installAsync$2$1.invokeSuspend(Unknown Source:30)
	at q3.a.resumeWith(Unknown Source:8)
	at e4.f0.run(Unknown Source:86)
	at h4.a.q(Unknown Source:0)
	at h4.a$a.run(:5)
```
Device information:
| key | value |
| --- | --- |
| version name | 75.4.0 |
| version code | 98 |
| build type | release |
| SDK version | 31 |
| android release | 12 |
| board | lmi |
| brand | POCO |
| device | lmi |
| manufacturer | Xiaomi |
| model | POCO F2 Pro |
| product | lmi_eea |
| SOC manufacturer | QTI |
| SOC model | SM8250 |
| supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| tags | release-keys |
| build time | 1649256271000 |

The problems continue.  After downloading the APK file there are times when the installation does not start.  I click on the option that gives me to go to the path of the cache folder and the APK file does not exist.  I have to download the file again one or more times until it finally downloads correctly and I can install.  Now it happens in both brave and Firefox.

**Comment by Tobi823** [comment]
I can't reproduce this bug on my devices and therefore have to poke around in the dark.

Can you try this version? I try to force the file creation with File::createNewFile() and BufferedOutputStream::flush.
[ffupdater-release.zip](https://github.com/Tobi823/ffupdater/files/8565363/ffupdater-release.zip)

**Comment by PsykoRider** [comment]
I have tried the new APK and I still have problems but now it does not show me any error code.  It only shows the window that sends me to the cache folder
![Screenshot_2022-04-26-19-15-31-481_de marmaro krt ffupdater](https://user-images.githubusercontent.com/87679784/165357325-70cb1524-bfea-4238-8e1e-6bc02bc7bc7d.jpg)
![Screenshot_2022-04-26-19-14-01-295_de marmaro krt ffupdater](https://user-images.githubusercontent.com/87679784/165357333-c2520341-75e0-4300-879e-a7d19c72ce00.jpg)
![Screenshot_2022-04-26-19-14-53-523_de marmaro krt ffupdater](https://user-images.githubusercontent.com/87679784/165357335-502ed73f-d491-465c-9f33-ee0a814d89f0.jpg)
![Screenshot_2022-04-26-19-14-59-698_de marmaro krt ffupdater](https://user-images.githubusercontent.com/87679784/165357340-2ad6d0f5-e793-4fe0-bfd5-869dc0ba4881.jpg)
![Screenshot_2022-04-26-19-15-14-321_de marmaro krt ffupdater](https://user-images.githubusercontent.com/87679784/165357341-34e24b97-3484-45b4-bce5-aab15cba7530.jpg)

**Comment by Tobi823** [comment]
Can you disable the MIUI optimization (settings in the Xiaomi Android operating system)? Maybe it helps.

**Comment by PsykoRider** [comment]
I would prefer not to disable miui optimizations because a while ago I disabled it and I had a lot of problems but after enabling it again the system recovered.  If you can't do anything else to figure out the problem, I apologize.

**Comment by Tobi823** [comment]
> I would prefer not to disable miui optimizations because a while ago I disabled it and I had a lot of problems but after enabling it again the system recovered.

Ok. I thought that disabling it is a painless process.

I will still think about what the error could be.

**Comment by Tobi823** [comment]
Maybe the "PackageInstaller" is buggy on Xiaomi and the old "Intent.ACTION_INSTALL_PACKAGE" ... can you try this newer version [ffupdater-release.zip](https://github.com/Tobi823/ffupdater/files/8565963/ffupdater-release.zip)?

**Comment by Tobi823** [comment]
Can other applications from the F-Droid (e.g. the AuroraStore) install apps? Maybe I can look in their source code and check what they do different.

**Comment by PsykoRider** [comment]
It seems that with this new APK it downloads and installs correctly.  I have tried it 4 times with brave and Firefox and it works correctly.  Thank you very much sir.  And yes, I can install APK files with Aurora store and other apps without problems.  If you need me to send you something to further address this error just ask.

**Comment by Tobi823** [comment]
> It seems that with this new APK it downloads and installs correctly.

Finally - This is good news 🎉

In the next offical FFUpdater version I will add an option to use the older  Intent.ACTION_INSTALL_PACKAGE-installer. You just have to enable the option in the settings.

**Comment by Tobi823** [comment]
Fix will be available in version 75.4.1
